noir-projects/aztec-nr/aztec/src/macros/dispatch/mod.nr

```diff
@@ -0,0 +1,141 @@
+pub comptime fn generate_public_dispatch(m: Module) -> Quoted {
+    let functions = m.functions();
+    let functions = functions.filter(|function: FunctionDefinition| function.has_named_attribute("public"));
+
+    let unit = get_type::<()>();
+
+    let ifs = functions.map(
+        |function: FunctionDefinition| {
+        let name = function.name();
+        let parameters = function.parameters();
+        let return_type = function.return_type();
+
+        let selector: Field = compute_fn_selector(function);
+
+        let mut parameters_size = 0;
+        for param in parameters {
+            parameters_size += size_in_fields(param.1);
+        }
+
+        let initial_read = if parameters.len() == 0 {
+            quote {}
+        } else {
+            // The initial calldata_copy offset is 1 to skip the Field selector
+            quote {
+                let input_calldata: [Field; $parameters_size] = dep::aztec::context::public_context::calldata_copy(1, $parameters_size);
+                let mut reader = dep::aztec::protocol_types::utils::reader::Reader::new(input_calldata);
+            }
+        };
+
+        let mut parameter_index = 0;
+        let reads = parameters.map(|param: (Quoted, Type)| {
+            let param_name = f"arg{parameter_index}".quoted_contents();
+            let param_type = param.1;
+            let read = quote {
+                let $param_name: $param_type = reader.read_struct(dep::aztec::protocol_types::traits::Deserialize::deserialize);
+            };
+            parameter_index += 1;
+            quote { $read }
+        });
+        let read = reads.join(quote { });
+
+        let mut args = &[];
+        for parameter_index in 0..parameters.len() {
+            let param_name = f"arg{parameter_index}".quoted_contents();
+            args = args.push_back(quote { $param_name });
+        }
+
+        let args = args.join(quote { , });
+        let call = quote { $name($args) };
+
+        let return_code = if return_type == unit {
+            quote { $call }
+        } else {
+            quote {
+                let return_value = dep::aztec::protocol_types::traits::Serialize::serialize($call);
+                dep::aztec::context::public_context::avm_return(return_value);
+            }
+        };
+
+        let if_ = quote {
+            if selector == $selector {
+                $initial_read
+                $read
+                $return_code
+            }
+        };
+        if_
+    }
+    );
+
+    if ifs.len() == 0 {
+        // No dispatch function if there are no public functions
+        quote {}
+    } else {
+        let ifs = ifs.push_back(quote { { panic(f"Unknown selector") } });
+        let dispatch = ifs.join(quote { else });
+
+        let body = quote {
+            unconstrained pub fn public_dispatch(selector: Field) {
+                $dispatch
+            }
+        };
+
+        body
+    }
+}
+
+comptime fn size_in_fields(typ: Type) -> u32 {
```

Comment:
I'm stealing this util for other macros instead of relying on `flatten_to_fields` 😁